internal/server/ingest/ingest.go

```diff
@@ -0,0 +1,229 @@
+//
+// It watches the configured location for new reports that have been saved to disk.
+// It will then process, validate, and then upload the reports to the database.
+func (s *Service) Run() error {
+	slog.Info("Ingest service started")
+
+	select {
+	case <-s.gracefulCtx.Done():
+		return errors.New("server is already shutting down")
+	default:
+	}
+
+	reloadEventCh, cfgWatchErrCh, err := s.cm.Watch(s.gracefulCtx)
+	if err != nil {
+		return fmt.Errorf("failed to start watch configuration: %v", err)
+	}
+
+	// 2) initial sync
+	s.syncWorkers()
+
+	// Debounce timer for handling bursts of events
+	debounceDuration := 5 * time.Second
+	debounceTimer := time.NewTimer(debounceDuration)
+	defer debounceTimer.Stop()
+
+	for {
+		select {
+		case <-s.gracefulCtx.Done():
+			slog.Info("Ingest service stopped")
+			return nil
+
+		case _, ok := <-reloadEventCh:
+			if !ok {
+				return fmt.Errorf("reloadEventCh closed unexpectedly")
+			}
+			if !debounceTimer.Stop() {
+				select {
+				case <-debounceTimer.C:
+				default:
+				}
+			}
+			debounceTimer.Reset(debounceDuration)
+
+		case <-debounceTimer.C:
+			// Timer expired, perform the resync
+			slog.Info("Resyncing workers after configuration change")
+			s.syncWorkers()
+			slog.Debug("Completed resyncing workers")
+
+		case err, ok := <-cfgWatchErrCh:
+			if !ok {
+				return fmt.Errorf("cfgWatchErrCh closed unexpectedly")
+			}
+			if err != nil {
+				slog.Error("Configuration watcher error", "err", err)
+			}
+		}
+	}
+}
+
+// syncWorkers diffs the allow‐list and starts/stops goroutines.
+func (s *Service) syncWorkers() {
+	allowed := s.cm.AllowList()
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	want := map[string]struct{}{}
+	for _, app := range allowed {
+		want[app] = struct{}{}
+	}
+
+	// stop removed
+	for app, cancel := range s.workers {
+		if _, ok := want[app]; !ok {
+			cancel()
+			delete(s.workers, app)
+		}
+	}
+	// start added
+	for app := range want {
+		if _, ok := s.workers[app]; ok {
+			continue
+		}
+
+		select {
+		case <-s.gracefulCtx.Done():
+			slog.Info("Graceful shutdown in progress, stopping app worker", "app", app)
+			return // normal shutdown
+		default:
+		}
+		ctx, cancel := context.WithCancel(s.gracefulCtx)
+		s.workers[app] = cancel
+		slog.Info("Starting app worker", "app", app)
+		go s.appWorker(ctx, app)
+	}
+}
+
+// appWorker watches & processes files for a single app until ctx is canceled.
+func (s *Service) appWorker(ctx context.Context, app string) {
+	inputDir := filepath.Join(s.cm.BaseDir(), app)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// this will read/process/remove JSON files and call s.db.Upload(...)
+			err := processor.ProcessFiles(ctx, inputDir, s.db)
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					slog.Info("Graceful shutdown in progress, stopping app worker", "app", app)
+					return // normal shutdown
+				}
+				slog.Error("Failed to process files", "app", app, "err", err)
+			}
+		}
+	}
```

Comment:
What happens if the call to `getJSONFiles()` inside `processor.ProcessFiles()` return an error? IIUC the error won't be `context.Cancelled` nor the context will be cancelled because of that, so this will spin forever and could even cause CPU thrashing, couldn't it?